File: scripts/converge.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "fileutils"
require "shellwords"
require "time"

def usage
  puts <<~EOF
    converge.rb - run rotating agent loop with optional session artifacts and handoff

    USAGE
      converge.rb [--session-dir <path>] --prompt-list <path> --agent-cmd "<command>" [--agent-cmd "<command>" ...] [--max-steps <n>] [--tmux] [--tmux-session-name <name>] [--handoff | --no-handoff]

    REQUIRED
      --prompt-list   File with one prompt path per line (rotation order).
      --agent-cmd     Agent command, e.g. "codex exec" or "claude".
                      Repeat to rotate commands independently from prompts.

    OPTIONAL
      --session-dir   Session root for run artifacts.
      --max-steps     Number of loop iterations (default: 10).
      --tmux          Run each step in a tmux window for live observability.
      --tmux-session-name
                      Optional tmux session name override.
      --handoff       Force handoff artifacts on. Requires --session-dir.
      --no-handoff    Disable handoff artifacts.
      -h, --help      Show this help message.

    PROMPT LIST RULES
      - Empty lines are ignored.
      - Lines starting with # are ignored.
      - Relative prompt paths are resolved from the prompt-list directory.

    EXAMPLES
      ruby scripts/converge.rb --prompt-list ./prompts.txt --agent-cmd "codex exec --dangerously-bypass-approvals-and-sandbox -" --max-steps 2
      ruby scripts/converge.rb --session-dir ./session --prompt-list ./prompts.txt --agent-cmd "claude -p --permission-mode bypassPermissions" --no-handoff
      ruby scripts/converge.rb --session-dir ./session --prompt-list ./prompts.txt --agent-cmd "cursor-agent -p --yolo --trust --approve-mcps" --handoff
  EOF
end

def iso_now
  Time.now.utc.strftime("%Y-%m-%dT%H:%M:%SZ")
end

def sanitize_tmux_label(text, fallback, max_len)
  safe = text.to_s.strip.gsub(/[^A-Za-z0-9._-]+/, "-").gsub(/-+/, "-").gsub(/\A-+|-+\z/, "")
  safe = safe[0, max_len]
  safe.nil? || safe.empty? ? fallback : safe
end

def build_tmux_session_name(provided)
  return provided unless provided.to_s.empty?

  "converge-#{Time.now.utc.strftime("%Y%m%d-%H%M%S")}-#{Process.pid}"
end

def build_tmux_window_name(step, prompt)
  suffix = sanitize_tmux_label(File.basename(prompt, File.extname(prompt)), "step", 24)
  format("s%03d-%s", step, suffix)
end

def tmux_available?
  ENV.fetch("PATH", "").split(File::PATH_SEPARATOR).any? do |dir|
    File.executable?(File.join(dir, "tmux"))
  end
end

def configure_tmux_window(session_name, window_name)
  target = "#{session_name}:#{window_name}"
  system("tmux", "set-option", "-t", target, "remain-on-exit", "on") or exit 1
  system("tmux", "set-option", "-t", target, "automatic-rename", "off") or exit 1
end

def build_tmux_step_command(payload, agent_cmd:, stdout_log: nil, stderr_log: nil)
  command = "printf '%s' #{Shellwords.escape(payload)} | bash -c #{Shellwords.escape(agent_cmd)}"
  if stdout_log && stderr_log
    command += " > >(tee #{Shellwords.escape(stdout_log)}) 2> >(tee #{Shellwords.escape(stderr_log)} >&2)"
  end
  "set -o pipefail; #{command}"
end

def wait_for_tmux_window(session_name, window_name)
  target = "#{session_name}:#{window_name}"
  loop do
    dead = `tmux list-panes -t #{Shellwords.escape(target)} -F '\#{pane_dead}' 2>/dev/null`.strip
    break if dead == "1"

    sleep 0.1
  end
  code = `tmux display-message -p -t #{Shellwords.escape(target)} '\#{pane_dead_status}'`.strip
  Integer(code.empty? ? "1" : code)
end

def build_effective_prompt(step, agent_cmd:, prompt:, input_handoff: nil, output_handoff: nil)
  header = []
  header << "# Runtime Protocol"
  header << ""
  header << "- step: #{step}"
  header << "- agent_cmd: #{agent_cmd}"
  if output_handoff
    if input_handoff
      header << "- input_handoff: #{input_handoff}"
      header << "- read input handoff as latest context."
    else
      header << "- input_handoff: (none)"
      header << "- no previous handoff exists for this step."
    end
    header << "- output_handoff: #{output_handoff}"
    header << "- write next handoff content to output_handoff."
  end
  header << "- do not modify files outside the task scope."
  header << ""
  header << "# Role Prompt"
  header << ""
  header.join("\n") + File.read(prompt)
end

def require_value!(args, index, option, allow_empty: true)
  value = args[index + 1]
  if value.nil?
    warn "Missing value for #{option}"
    usage
    exit 1
  end

  if !allow_empty && value.empty?
    warn "Invalid value for #{option}"
    usage
    exit 1
  end

  value
end

opts = { max_steps: 10, tmux: false, agent_cmds: [], handoff: nil }
i = 0
while i < ARGV.length
  case ARGV[i]
  when "--session-dir" then opts[:session_dir] = require_value!(ARGV, i, "--session-dir"); i += 2
  when "--prompt-list" then opts[:prompt_list] = require_value!(ARGV, i, "--prompt-list"); i += 2
  when "--agent-cmd" then opts[:agent_cmds] << require_value!(ARGV, i, "--agent-cmd", allow_empty: false); i += 2
  when "--max-steps" then opts[:max_steps] = Integer(require_value!(ARGV, i, "--max-steps")); i += 2
  when "--tmux" then opts[:tmux] = true; i += 1
  when "--tmux-session-name" then opts[:tmux_session_name] = require_value!(ARGV, i, "--tmux-session-name"); i += 2
  when "--handoff" then opts[:handoff] = true; i += 1
  when "--no-handoff" then opts[:handoff] = false; i += 1
  when "-h", "--help" then usage; exit 0
  else
    warn "Unknown argument: #{ARGV[i]}"
    usage
    exit 1
  end
end

if opts[:prompt_list].to_s.empty? || opts[:agent_cmds].empty?
  usage
  exit 1
end
if opts[:max_steps] < 1
  warn "--max-steps must be a positive integer."
  exit 1
end
session_dir = opts[:session_dir].to_s.empty? ? nil : File.expand_path(opts[:session_dir])
if opts[:handoff] && session_dir.nil?
  warn "--handoff requires --session-dir."
  exit 1
end
handoff_enabled = opts[:handoff].nil? ? !session_dir.nil? : opts[:handoff]

FileUtils.mkdir_p(session_dir) if session_dir
prompt_list = File.expand_path(opts[:prompt_list])
unless File.file?(prompt_list)
  warn "Prompt list not found: #{prompt_list}"
  exit 1
end

prompt_base = File.dirname(prompt_list)
prompts = []
File.readlines(prompt_list, chomp: true).each do |raw|
  line = raw.rstrip
  next if line.empty? || line.start_with?("#")

  path = line.start_with?("/") ? line : File.join(prompt_base, line)
  path = File.expand_path(path)
  unless File.file?(path)
    warn "Prompt file not found: #{path}"
    exit 1
  end
  prompts << path
end
if prompts.empty?
  warn "Prompt list contains no usable prompt files."
  exit 1
end

run_dir = nil
loop_log = nil
if session_dir
  run_dir = File.join(session_dir, "run")
  loop_dir = File.join(run_dir, "loop")
  loop_log = File.join(loop_dir, "loop.log")
  FileUtils.mkdir_p(loop_dir)
  FileUtils.touch(loop_log)
end

tmux_session_name = nil
tmux_created = false
if opts[:tmux]
  unless tmux_available?
    warn "--tmux requires tmux on PATH."
    exit 1
  end
  tmux_session_name = build_tmux_session_name(opts[:tmux_session_name])
  if system("tmux", "has-session", "-t", tmux_session_name, out: File::NULL, err: File::NULL)
    warn "tmux session already exists: #{tmux_session_name}"
    exit 1
  end
end

puts "Starting agent loop"
puts "session_dir=#{session_dir}" if session_dir
puts "prompt_count=#{prompts.length}"
puts "max_steps=#{opts[:max_steps]}"
if opts[:agent_cmds].length == 1
  puts "agent_cmd=#{opts[:agent_cmds].first}"
else
  puts "agent_cmd_count=#{opts[:agent_cmds].length}"
end
if tmux_session_name
  puts "tmux_session=#{tmux_session_name}"
  puts "tmux_attach_cmd=tmux attach -t #{Shellwords.escape(tmux_session_name)}"
end

1.upto(opts[:max_steps]) do |step|
  prompt = prompts[(step - 1) % prompts.length]
  agent_cmd = opts[:agent_cmds][(step - 1) % opts[:agent_cmds].length]

  input_handoff = nil
  output_handoff = nil
  step_dir = nil
  stdout_log = nil
  stderr_log = nil
  code = nil

  started = Time.now.to_i
  puts "[step #{step}] start prompt=#{File.basename(prompt)} time=#{iso_now}"

  if session_dir
    step_dir = File.join(run_dir, format("s%03d", step))
    FileUtils.mkdir_p(step_dir)
    effective = File.join(step_dir, "effective_prompt.md")
    stdout_log = File.join(step_dir, "stdout.log")
    stderr_log = File.join(step_dir, "stderr.log")
    File.write(stdout_log, "")
    File.write(stderr_log, "")
    if handoff_enabled
      if step > 1
        prev = File.join(run_dir, format("s%03d", step - 1), "handoff.md")
        input_handoff = prev if File.file?(prev)
      end
      output_handoff = File.join(step_dir, "handoff.md")
      File.write(output_handoff, "")
    end
    File.write(
      effective,
      build_effective_prompt(
        step,
        agent_cmd: agent_cmd,
        prompt: prompt,
        input_handoff: input_handoff,
        output_handoff: output_handoff
      )
    )
    unless opts[:tmux]
      File.open(effective, "r") do |input|
        File.open(stdout_log, "w") do |output|
          File.open(stderr_log, "w") do |error|
            system("bash", "-c", agent_cmd, in: input, out: output, err: error)
            code = $?.exitstatus
            File.write(File.join(step_dir, "exit_code.txt"), "#{code}\n")
          end
        end
      end
    end
  end

  if opts[:tmux]
    tmux_cmd = build_tmux_step_command(
      build_effective_prompt(
        step,
        agent_cmd: agent_cmd,
        prompt: prompt,
        input_handoff: input_handoff,
        output_handoff: output_handoff
      ),
      agent_cmd: agent_cmd,
      stdout_log: stdout_log,
      stderr_log: stderr_log
    )
    window_name = build_tmux_window_name(step, prompt)
    if !tmux_created
      system("tmux", "new-session", "-d", "-s", tmux_session_name, "-n", window_name, "bash", "-c", tmux_cmd) or exit 1
      configure_tmux_window(tmux_session_name, window_name)
      tmux_created = true
    else
      system("tmux", "new-window", "-d", "-t", tmux_session_name, "-n", window_name, "bash", "-c", tmux_cmd) or exit 1
      configure_tmux_window(tmux_session_name, window_name)
    end
    code = wait_for_tmux_window(tmux_session_name, window_name)
    File.write(File.join(step_dir, "exit_code.txt"), "#{code}\n") if step_dir
  elsif !session_dir
    reader, writer = IO.pipe
    writer_thread = Thread.new do
      begin
        writer.write(build_effective_prompt(step, agent_cmd: agent_cmd, prompt: prompt))
      rescue Errno::EPIPE
        nil
      ensure
        writer.close
      end
    end
    system("bash", "-c", agent_cmd, in: reader)
    code = $?.exitstatus
    reader.close
    writer_thread.join
  end

  elapsed = Time.now.to_i - started
  puts "[step #{step}] done exit=#{code} elapsed=#{elapsed}s"
  if loop_log
    File.open(loop_log, "a") do |f|
      f.puts("#{iso_now} step=#{step} prompt=#{prompt} agent_cmd=#{Shellwords.escape(agent_cmd)} exit=#{code} elapsed_s=#{elapsed}")
    end
  end
end

puts "Loop finished after #{opts[:max_steps]} steps."

File: scripts/converge.py
#!/usr/bin/env python3
import argparse
import os
import re
import shlex
import shutil
import subprocess
import sys
import time
from datetime import datetime, timezone
from pathlib import Path


def iso_now() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def sanitize_tmux_label(text: str, fallback: str, max_len: int) -> str:
    safe = re.sub(r"[^A-Za-z0-9._-]+", "-", text.strip())
    safe = re.sub(r"-{2,}", "-", safe).strip("-")
    safe = safe[:max_len]
    return safe or fallback


def build_tmux_session_name(provided: str | None) -> str:
    if provided:
        return provided
    stamp = datetime.now(timezone.utc).strftime("%Y%m%d-%H%M%S")
    return f"converge-{stamp}-{os.getpid()}"


def build_tmux_window_name(step: int, prompt: Path) -> str:
    suffix = sanitize_tmux_label(prompt.stem, "step", 24)
    return f"s{step:03d}-{suffix}"


def shell_escape_arg(text: str) -> str:
    escaped = re.sub(r"([^A-Za-z0-9_\-.,:+/@\n])", r"\\\1", text)
    return escaped.replace("\n", "'\n'")


def build_tmux_step_command(
    payload: str,
    *,
    agent_cmd: str,
    stdout_log: Path | None = None,
    stderr_log: Path | None = None,
) -> str:
    q = shlex.quote
    command = f"printf '%s' {q(payload)} | bash -c {q(agent_cmd)}"
    if stdout_log is not None and stderr_log is not None:
        command += f" > >(tee {q(str(stdout_log))}) 2> >(tee {q(str(stderr_log))} >&2)"
    return f"set -o pipefail; {command}"


def wait_for_tmux_window(session_name: str, window_name: str) -> int:
    target = f"{session_name}:{window_name}"
    while True:
        proc = subprocess.run(
            ["tmux", "list-panes", "-t", target, "-F", "#{pane_dead}"],
            check=False,
            capture_output=True,
            text=True,
        )
        if proc.returncode == 0 and proc.stdout.strip() == "1":
            break
        time.sleep(0.1)

    status = subprocess.run(
        ["tmux", "display-message", "-p", "-t", target, "#{pane_dead_status}"],
        check=True,
        capture_output=True,
        text=True,
    )
    raw = status.stdout.strip()
    return int(raw or "1")


def configure_tmux_window(session_name: str, window_name: str) -> None:
    target = f"{session_name}:{window_name}"
    subprocess.run(["tmux", "set-option", "-t", target, "remain-on-exit", "on"], check=True)
    subprocess.run(["tmux", "set-option", "-t", target, "automatic-rename", "off"], check=True)


def resolve_prompts(prompt_list: Path) -> list[Path]:
    prompts: list[Path] = []
    base = prompt_list.parent
    for raw in prompt_list.read_text().splitlines():
        line = raw.rstrip()
        if not line or line.startswith("#"):
            continue
        p = Path(line)
        if not p.is_absolute():
            p = base / p
        p = p.resolve()
        if not p.is_file():
            raise FileNotFoundError(f"Prompt file not found: {p}")
        prompts.append(p)
    if not prompts:
        raise ValueError("Prompt list contains no usable prompt files.")
    return prompts


def build_effective_prompt(
    step: int,
    *,
    agent_cmd: str,
    prompt: Path,
    input_handoff: Path | None = None,
    output_handoff: Path | None = None,
) -> str:
    header = [
        "# Runtime Protocol",
        "",
        f"- step: {step}",
        f"- agent_cmd: {agent_cmd}",
    ]
    if output_handoff is not None:
        header.extend(
            [
                f"- input_handoff: {input_handoff}" if input_handoff else "- input_handoff: (none)",
                "- read input handoff as latest context."
                if input_handoff
                else "- no previous handoff exists for this step.",
                f"- output_handoff: {output_handoff}",
                "- write next handoff content to output_handoff.",
            ]
        )
    header.extend(
        [
            "- do not modify files outside the task scope.",
            "",
            "# Role Prompt",
            "",
        ]
    )
    return "\n".join(header) + prompt.read_text()


def main() -> int:
    ap = argparse.ArgumentParser(
        prog="converge.py",
        description="Run a rotating agent loop with optional session artifacts and handoff.",
        formatter_class=argparse.RawTextHelpFormatter,
        epilog=(
            "Prompt list rules:\n"
            "  - empty lines are ignored\n"
            "  - lines starting with # are ignored\n"
            "  - relative prompt paths are resolved from prompt-list directory\n\n"
            "Examples:\n"
            "  python3 scripts/converge.py --prompt-list ./prompts.txt --agent-cmd \"codex exec --dangerously-bypass-approvals-and-sandbox -\" --max-steps 2\n"
            "  python3 scripts/converge.py --session-dir ./session --prompt-list ./prompts.txt --agent-cmd \"claude -p --permission-mode bypassPermissions\" --no-handoff\n"
            "  python3 scripts/converge.py --session-dir ./session --prompt-list ./prompts.txt --agent-cmd \"cursor-agent -p --yolo --trust --approve-mcps\" --handoff"
        ),
    )
    ap.add_argument("--session-dir", help="Optional session root for run artifacts.")
    ap.add_argument("--prompt-list", required=True, help="File with one prompt path per line.")
    ap.add_argument(
        "--agent-cmd",
        action="append",
        required=True,
        help="Agent command, e.g. 'codex exec' or 'claude'. Repeat to rotate commands by step.",
    )
    ap.add_argument("--max-steps", type=int, default=10, help="Number of loop iterations (default: 10).")
    ap.add_argument("--tmux", action="store_true", help="Run each step in a tmux window for live observability.")
    ap.add_argument("--tmux-session-name", help="Optional tmux session name override.")
    handoff = ap.add_mutually_exclusive_group()
    handoff.add_argument("--handoff", dest="handoff", action="store_true", help="Force handoff artifacts on. Requires --session-dir.")
    handoff.add_argument("--no-handoff", dest="handoff", action="store_false", help="Disable handoff artifacts.")
    ap.set_defaults(handoff=None)
    args = ap.parse_args()
    if args.max_steps < 1:
        print("--max-steps must be a positive integer.", file=sys.stderr)
        return 1

    session_dir: Path | None = None
    if args.session_dir:
        session_dir = Path(args.session_dir).resolve()
        session_dir.mkdir(parents=True, exist_ok=True)
    if args.handoff and session_dir is None:
        print("--handoff requires --session-dir.", file=sys.stderr)
        return 1
    prompt_list = Path(args.prompt_list).resolve()
    if not prompt_list.is_file():
        print(f"Prompt list not found: {prompt_list}", file=sys.stderr)
        return 1
    prompts = resolve_prompts(prompt_list)
    agent_cmds: list[str] = args.agent_cmd
    handoff_enabled = args.handoff if args.handoff is not None else session_dir is not None

    run_dir: Path | None = None
    loop_log: Path | None = None
    if session_dir is not None:
        run_dir = session_dir / "run"
        loop_dir = run_dir / "loop"
        loop_dir.mkdir(parents=True, exist_ok=True)
        loop_log = loop_dir / "loop.log"
        loop_log.touch(exist_ok=True)

    tmux_session_name: str | None = None
    tmux_created = False
    if args.tmux:
        if shutil.which("tmux") is None:
            print("--tmux requires tmux on PATH.", file=sys.stderr)
            return 1
        tmux_session_name = build_tmux_session_name(args.tmux_session_name)
        has_session = subprocess.run(
            ["tmux", "has-session", "-t", tmux_session_name],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            check=False,
        )
        if has_session.returncode == 0:
            print(f"tmux session already exists: {tmux_session_name}", file=sys.stderr)
            return 1

    print("Starting agent loop")
    if session_dir is not None:
        print(f"session_dir={session_dir}")
    print(f"prompt_count={len(prompts)}")
    print(f"max_steps={args.max_steps}")
    if len(agent_cmds) == 1:
        print(f"agent_cmd={agent_cmds[0]}")
    else:
        print(f"agent_cmd_count={len(agent_cmds)}")
    if tmux_session_name:
        print(f"tmux_session={tmux_session_name}")
        print(f"tmux_attach_cmd=tmux attach -t {shell_escape_arg(tmux_session_name)}")

    for step in range(1, args.max_steps + 1):
        prompt = prompts[(step - 1) % len(prompts)]
        agent_cmd = agent_cmds[(step - 1) % len(agent_cmds)]
        input_handoff: Path | None = None
        output_handoff: Path | None = None
        effective = stdout_log = stderr_log = None
        step_dir: Path | None = None
        if session_dir is not None and run_dir is not None:
            step_dir = run_dir / f"s{step:03d}"
            step_dir.mkdir(parents=True, exist_ok=True)
            stdout_log = step_dir / "stdout.log"
            stderr_log = step_dir / "stderr.log"
            stdout_log.write_text("")
            stderr_log.write_text("")
            if handoff_enabled:
                prev_handoff = run_dir / f"s{step - 1:03d}" / "handoff.md"
                input_handoff = prev_handoff if step > 1 and prev_handoff.is_file() else None
                output_handoff = step_dir / "handoff.md"
                output_handoff.write_text("")
            effective = step_dir / "effective_prompt.md"
            effective.write_text(
                build_effective_prompt(
                    step,
                    agent_cmd=agent_cmd,
                    prompt=prompt,
                    input_handoff=input_handoff,
                    output_handoff=output_handoff,
                )
            )

        start_epoch = int(time.time())
        start_iso = iso_now()
        print(f"[step {step}] start prompt={prompt.name} time={start_iso}")
        if tmux_session_name:
            payload = build_effective_prompt(
                step,
                agent_cmd=agent_cmd,
                prompt=prompt,
                input_handoff=input_handoff,
                output_handoff=output_handoff,
            )
            window_name = build_tmux_window_name(step, prompt)
            tmux_cmd = build_tmux_step_command(
                payload,
                agent_cmd=agent_cmd,
                stdout_log=stdout_log,
                stderr_log=stderr_log,
            )
            if not tmux_created:
                subprocess.run(
                    ["tmux", "new-session", "-d", "-s", tmux_session_name, "-n", window_name, "bash", "-c", tmux_cmd],
                    check=True,
                )
                configure_tmux_window(tmux_session_name, window_name)
                tmux_created = True
            else:
                subprocess.run(
                    ["tmux", "new-window", "-d", "-t", tmux_session_name, "-n", window_name, "bash", "-c", tmux_cmd],
                    check=True,
                )
                configure_tmux_window(tmux_session_name, window_name)
            code = wait_for_tmux_window(tmux_session_name, window_name)
            if step_dir is not None:
                (step_dir / "exit_code.txt").write_text(f"{code}\n")
        elif session_dir is not None:
            assert effective is not None
            assert stdout_log is not None
            assert stderr_log is not None
            with effective.open("r") as input_stream, stdout_log.open("w") as out, stderr_log.open("w") as err:
                proc = subprocess.run(["bash", "-c", agent_cmd], text=True, stdin=input_stream, stdout=out, stderr=err)
            code = proc.returncode
            (step_dir / "exit_code.txt").write_text(f"{code}\n")
        else:
            proc = subprocess.run(
                ["bash", "-c", agent_cmd],
                text=True,
                input=build_effective_prompt(step, agent_cmd=agent_cmd, prompt=prompt),
            )
            code = proc.returncode

        end_iso = iso_now()
        elapsed = int(time.time()) - start_epoch
        print(f"[step {step}] done exit={code} elapsed={elapsed}s")
        if loop_log is not None:
            with loop_log.open("a") as lf:
                lf.write(
                    f"{end_iso} step={step} prompt={prompt} agent_cmd={shlex.quote(agent_cmd)} exit={code} elapsed_s={elapsed}\n"
                )

    print(f"Loop finished after {args.max_steps} steps.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
